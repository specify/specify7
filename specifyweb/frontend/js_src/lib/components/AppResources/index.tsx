import React from 'react';

import { useCachedState } from '../../hooks/useCachedState';
import { useErrorContext } from '../../hooks/useErrorContext';
import { resourcesText } from '../../localization/resources';
import type { GetOrSet } from '../../utils/types';
import { Container, H2 } from '../Atoms';
import { ProtectedTable, ProtectedTool } from '../Permissions/PermissionDenied';
import { SafeOutlet } from '../Router/RouterUtils';
import { AppResourcesAside } from './Aside';
import { AppResourcesFilters } from './Filters';
import type { AppResources } from './hooks';
import { useAppResources } from './hooks';

export function AppResourcesWrapper(): JSX.Element {
  return (
    <ProtectedTool action="read" tool="resources">
      <ProtectedTable action="read" tableName="Discipline">
        <ProtectedTable action="read" tableName="Collection">
          <ProtectedTable action="read" tableName="SpecifyUser">
            <AppResourcesDataFetcher />
          </ProtectedTable>
        </ProtectedTable>
      </ProtectedTable>
    </ProtectedTool>
  );
}

function AppResourcesDataFetcher(): JSX.Element | null {
  const getSetResources = useAppResources();
  const [resources] = getSetResources;
  useErrorContext('appResourcesData', resources);

  return typeof resources === 'object' ? (
    <AppResourcesView getSet={getSetResources as GetOrSet<AppResources>} />
  ) : null;
}

export type AppResourcesOutlet = { readonly getSet: GetOrSet<AppResources> };

function AppResourcesView({
  getSet,
}: {
  readonly getSet: GetOrSet<AppResources>;
}): JSX.Element {
  const [resources] = getSet;

  const conformations = useCachedState('appResources', 'conformation');
  const [filters] = useCachedState('appResources', 'filters');

  return (
    <Container.FullGray>
      <div className="flex flex-wrap items-center gap-4">
        <H2 className="text-2xl">{resourcesText.resources()}</H2>
        <div className="flex gap-2">
          <AppResourcesFilters initialResources={resources} />
        </div>
      </div>
      <div className="flex flex-1 flex-col gap-4 sm:h-0 sm:overflow-scroll md:flex-row">
        <AppResourcesAside
          conformations={conformations}
          filters={filters}
          isEmbedded={false}
          resources={resources}
        />
        <SafeOutlet<AppResourcesOutlet>
          getSet={getSet as GetOrSet<AppResources>}
        />
      </div>
    </Container.FullGray>
  );
}
