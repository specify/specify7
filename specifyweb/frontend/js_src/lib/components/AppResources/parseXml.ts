export function strictParseXml(xml: string): Element {
  const parsed = parseXml(xml);
  // eslint-disable-next-line functional/no-throw-statement
  if (typeof parsed === 'string') throw new Error(parsed);
  else return parsed;
}

export function parseXml(string: string): Element | string {
  const parsedXml = new globalThis.DOMParser().parseFromString(
    string,
    'text/xml'
  ).documentElement;

  // Chrome, Safari
  const parseError = parsedXml.getElementsByTagName('parsererror')[0];
  if (typeof parseError === 'object')
    return (parseError.children[1].textContent ?? parseError.innerHTML).trim();
  // Firefox
  else if (parsedXml.tagName === 'parsererror')
    return (
      parsedXml.childNodes[0].nodeValue ??
      parsedXml.textContent ??
      parsedXml.innerHTML
    ).trim();
  else return parsedXml;
}
