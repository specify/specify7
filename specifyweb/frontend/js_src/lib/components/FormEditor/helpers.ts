import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
