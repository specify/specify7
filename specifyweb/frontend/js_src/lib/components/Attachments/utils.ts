import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
