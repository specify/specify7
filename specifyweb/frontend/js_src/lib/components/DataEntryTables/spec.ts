import { f } from '../../utils/functools';
import { pipe } from '../Syncer';
import { syncers } from '../Syncer/syncers';
import type { SimpleXmlNode } from '../Syncer/xmlToJson';
import { createSimpleXmlNode } from '../Syncer/xmlToJson';
import { createXmlSpec } from '../Syncer/xmlUtils';

export const dataEntryItems = f.store(() =>
  createXmlSpec({
    items: pipe(
      syncers.xmlChild('std'),
      syncers.fallback<SimpleXmlNode>(createSimpleXmlNode),
      syncers.xmlChildren('view'),
      syncers.map(syncers.object(dataEntryItem()))
    ),
  })
);

const dataEntryItem = f.store(() =>
  createXmlSpec({
    viewName: syncers.xmlAttribute('view', 'required'),
  })
);
