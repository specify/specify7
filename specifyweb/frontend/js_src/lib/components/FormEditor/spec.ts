import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import { defined, localized } from '../../utils/types';
import { getUniqueName } from '../../utils/uniquifyName';
import { strictParseXml } from '../AppResources/parseXml';
import type { ViewDefinition } from '../FormParse';
import type { SpecToJson } from '../Syncer';
import { pipe, syncer } from '../Syncer';
import { syncers } from '../Syncer/syncers';
import type { SimpleXmlNode, XmlNode } from '../Syncer/xmlToJson';
import {
  createSimpleXmlNode,
  toSimpleXmlNode,
  xmlToJson,
} from '../Syncer/xmlToJson';
import {
  createXmlSpec,
  getOriginalSyncerInput,
  setOriginalSyncerInput,
} from '../Syncer/xmlUtils';
import { getBusinessRuleClassFromTable } from './helpers';

export const viewSetsSpec = f.store(() =>
  createXmlSpec({
    views: pipe(
      syncers.xmlChild('views'),
      syncers.fallback(createSimpleXmlNode),
      syncers.xmlChildren('view'),
      syncers.map(resolvedViewSpec())
    ),
    viewDefs: pipe(
      syncers.xmlChild('viewdefs'),
      syncers.fallback(createSimpleXmlNode),
      syncers.xmlChildren('viewdef'),
      syncers.map(resolvedViewDefSpec())
    ),
    name: pipe(
      syncers.xmlAttribute('name', 'empty'),
      syncers.default<string>('')
    ),
  })
);

type RawViewSets = SpecToJson<ReturnType<typeof viewSetsSpec>>;
export type ViewSets = Omit<RawViewSets, 'viewDefs'> & {
  readonly viewDefs: RA<
    Omit<RawViewSets['viewDefs'][number], 'raw'> & { readonly raw: XmlNode }
  >;
};

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const resolvedViewSpec = () =>
  pipe(
    syncers.object(viewSpec()),
    syncer(
      ({ table, ...node }) => ({
        ...node,
        table: table.parsed,
        legacyTable: table.bad,
      }),
      ({ table, legacyTable, businessRules, ...node }) => ({
        ...node,
        table: { parsed: table, bad: legacyTable },
        businessRules:
          f.trim(businessRules ?? '') === ''
            ? localized(
                typeof table === 'object'
                  ? getBusinessRuleClassFromTable(table.name)
                  : ''
              )
            : businessRules,
      })
    )
  );

export function parseFormView(definition: ViewDefinition) {
  const view = resolvedViewSpec().serializer(
    toSimpleXmlNode(xmlToJson(strictParseXml(definition.view)))
  );
  const usedDefinitions = new Set(
    view.altViews.altViews.map(({ viewDef }) => viewDef)
  );
  return {
    view,
    viewDefinitions: Object.values(definition.viewdefs)
      .map((definition) =>
        resolvedViewDefSpec().serializer(
          toSimpleXmlNode(xmlToJson(strictParseXml(definition)))
        )
      )
      .filter(({ name }) =>
        /*
         * Only duplicate view definitions that were used in altviews. I.e,
         * if "ObjectAttachment" viewdef view was referenced by
         * "CollectionObjectAttachment Table" viewdef, don't create another
         * ObjectAttachment
         */
        usedDefinitions.has(name)
      ),
  };
}

const viewSpec = f.store(() =>
  createXmlSpec({
    name: syncers.xmlAttribute('name', 'required'),
    // Used for tooltips in sp6
    title: syncers.xmlAttribute('objTitle', 'skip'),
    description: pipe(
      syncers.xmlChild('desc', 'optional'),
      syncers.maybe(syncers.xmlContent)
    ),
    table: pipe(
      syncers.xmlAttribute('class', 'required'),
      syncers.preserveInvalid(syncers.javaClassName(false))
    ),
    businessRules: pipe(
      syncers.xmlAttribute('busrules', 'skip'),
      syncers.default(localized(''))
    ),
    altViews: pipe(
      syncers.xmlChild('altviews'),
      syncers.fallback<SimpleXmlNode>(createSimpleXmlNode),
      syncers.object(altViewsSpec())
    ),
    legacyIsInternal: pipe(
      syncers.xmlAttribute('isInternal', 'skip'),
      syncers.maybe(syncers.toBoolean)
    ),
    // Not used in the code, but specified in the xml often
    legacyIsExternal: pipe(
      syncers.xmlAttribute('isExternal', 'skip'),
      syncers.maybe(syncers.toBoolean)
    ),
    legacyUseBusinessRules: pipe(
      syncers.xmlAttribute('useDefBusRule', 'skip'),
      syncers.maybe(syncers.toBoolean),
      syncers.default<boolean>(true)
    ),
    legacyResourceLabels: pipe(
      syncers.xmlAttribute('resourceLabels', 'skip'),
      syncers.maybe(syncers.toBoolean)
    ),
  })
);

const altViewsSpec = f.store(() =>
  createXmlSpec({
    altViews: pipe(
      syncers.xmlChildren('altview'),
      syncers.map(
        pipe(
          syncers.object(altViewSpec()),
          // Use viewDef as name if name not provided (to match sp6 behavior)
          syncers.change(
            'name',
            ({ name }) => name,
            ({ name, viewDef }) => name ?? viewDef
          )
        )
      ),
      // Sp6 expects altView names to be unique
      syncer(f.id, (altViews) =>
        altViews.map(({ name, ...altView }, index) => ({
          ...altView,
          name: getUniqueName(
            name,
            altViews.slice(0, index).map(({ name }) => name)
          ),
        }))
      )
    ),
    legacySelector: syncers.xmlAttribute('selector', 'skip'),
    // Present in the sp6 source code, but never used for anything
    legacyDefaultMode: pipe(
      syncers.xmlAttribute('defaultMode', 'skip'),
      syncers.maybe(syncers.enum(['view', 'edit', 'search'] as const))
    ),
  })
);

const altViewSpec = f.store(() =>
  createXmlSpec({
    name: pipe(
      syncers.xmlAttribute('name', 'required'),
      syncers.default<string>('')
    ),
    viewDef: syncers.xmlAttribute('viewDef', 'required'),
    mode: pipe(
      syncers.xmlAttribute('mode', 'required'),
      syncers.maybe(syncers.enum(['edit', 'view', 'search'] as const)),
      syncers.fallback<'edit' | 'search' | 'view'>('view' as const)
    ),
    default: pipe(
      syncers.xmlAttribute('default', 'skip'),
      syncers.maybe(syncers.toBoolean),
      syncers.default<boolean>(false)
    ),
    legacyTitle: syncers.xmlAttribute('title', 'skip'),
    legacyLabel: syncers.xmlAttribute('label', 'skip'),
    legacyValidated: pipe(
      syncers.xmlAttribute('validated', 'skip'),
      syncers.maybe(syncers.toBoolean)
    ),
    legacySelectorValue: syncers.xmlAttribute('selector_value', 'skip', false),
  })
);

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const resolvedViewDefSpec = () =>
  pipe(
    syncers.object(viewDefSpec()),
    syncer(
      ({ table, ...node }) => ({
        ...node,
        table: table.parsed,
        legacyTable: table.bad,
      }),
      ({ table, legacyTable, ...node }) => {
        const result = {
          ...node,
          table: {
            parsed: table,
            bad: legacyTable,
          },
          legacyGetTable: localized(
            node.legacyGetTable ??
              (node.name?.endsWith('Search') === true
                ? 'edu.ku.brc.af.ui.forms.DataGetterForHashMap'
                : 'edu.ku.brc.af.ui.forms.DataGetterForObj')
          ),
          legacySetTable: localized(
            node.legacySetTable ??
              (node.name?.endsWith('Search') === true
                ? 'edu.ku.brc.af.ui.forms.DataSetterForHashMap'
                : 'edu.ku.brc.af.ui.forms.DataSetterForObj')
          ),
        };
        setOriginalSyncerInput(result, node.raw);
        return result;
      }
    )
  );

const viewDefSpec = f.store(() =>
  createXmlSpec({
    name: syncers.xmlAttribute('name', 'required'),
    table: pipe(
      syncers.xmlAttribute('class', 'required'),
      syncers.preserveInvalid(syncers.javaClassName(false))
    ),
    type: pipe(
      syncers.xmlAttribute('type', 'required'),
      syncers.maybe(
        syncers.enum(['form', 'formtable', 'iconview', 'rstable'] as const)
      ),
      /*
       * REFACTOR: try improving type inference to not require explicitly
       *   specifying generic
       */
      syncers.fallback<'form' | 'formtable' | 'iconview' | 'rstable'>('form')
    ),
    legacyGetTable: syncers.xmlAttribute('getTable', 'required'),
    legacySetTable: syncers.xmlAttribute('setTable', 'required'),
    legacyEditableDialog: pipe(
      syncers.xmlAttribute('editableDlg', 'skip'),
      syncers.maybe(syncers.toBoolean),
      syncers.default<boolean>(true)
    ),
    legacyUseResourceLabels: pipe(
      syncers.xmlAttribute('useResourceLabels', 'skip'),
      syncers.maybe(syncers.toBoolean)
    ),
    /*
     * Not parsing the rest of the form definition but leaving it as is so
     * as not to slow down the performance too much for big files.
     * Instead, the contents of the form definition will validated by
     * formDefinitionSpec() later on
     */
    raw: syncer<SimpleXmlNode, XmlNode>(
      (node) => ({
        ...defined(getOriginalSyncerInput(node), ''),
        attributes: {},
      }),
      toSimpleXmlNode
    ),
  })
);
