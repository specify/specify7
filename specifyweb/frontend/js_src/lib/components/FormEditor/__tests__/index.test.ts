import { requireContext } from '../../../tests/helpers';
import { removeKey, replaceItem } from '../../../utils/utils';
import { strictParseXml } from '../../AppResources/parseXml';
import { formatXmlNode } from '../../Syncer/formatXmlNode';
import { syncers } from '../../Syncer/syncers';
import type { XmlNode } from '../../Syncer/xmlToJson';
import { jsonToXml, toSimpleXmlNode, xmlToJson } from '../../Syncer/xmlToJson';
import { updateXml, xmlToString } from '../../Syncer/xmlToString';
import { exportsForTests } from '../index';
import type { ViewSets } from '../spec';
import { viewSetsSpec } from '../spec';
import { testFormDefinition } from './testFormDefinition';

const { injectRawXml } = exportsForTests;

requireContext();

const xml = `<?xml version="1.0" encoding="UTF-8"?>
<viewset name="Fish Views" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
<views>
  <view
    name="Accession"
    class="edu.ku.brc.specify.datamodel.Accession"
    busrules="edu.ku.brc.specify.datamodel.busrules.AccessionBusRules"
    resourcelabels="false"
  >
    <desc><![CDATA[The Accession form.]]></desc>
    <!-- Some comment -->
    <altviews>
      <altview name="Accession View" viewdef="Accession" mode="view" default="true"/>
      <altview name="Accession Edit" viewdef="Accession" mode="edit"/>
    </altviews>
  </view>
</views>
<viewdefs>
  <viewdef
    type="form"
    name="Accession"
    class="edu.ku.brc.specify.datamodel.Accession"
    gettable="edu.ku.brc.af.ui.forms.DataGetterForObj"
    settable="edu.ku.brc.af.ui.forms.DataSetterForObj"
  >${testFormDefinition}</viewdef>
</viewdefs>
</viewset>
`;

const process = (node: XmlNode): string =>
  xmlToString(jsonToXml(formatXmlNode(node)));

test('Can edit a form definition', () => {
  const { serializer, deserializer } = syncers.object(viewSetsSpec());
  const xmlNode = xmlToJson(strictParseXml(xml));
  const parsed = serializer(toSimpleXmlNode(xmlNode));
  const augmented = injectRawXml(parsed);
  const raw = augmented.viewDefs[0].raw;

  // Verify that view sets are parsed correctly
  const reconstructed = {
    ...augmented,
    viewDefs: replaceItem(
      augmented.viewDefs,
      0,
      removeKey(augmented.viewDefs[0], 'raw')
    ),
  };
  // Get rid of symbols
  expect(JSON.parse(JSON.stringify(reconstructed))).toMatchSnapshot();

  // Verify that view definition xml can be converted back to the same string
  const initialXml = process({ ...raw, attributes: {} });
  expect(initialXml).toBe(
    process(
      xmlToJson(strictParseXml(`<viewdef>${testFormDefinition}</viewdef>`))
    )
  );

  // Make a small edit
  const columnDefinitionIndex = raw.children.findIndex(
    (node) => node.type === 'XmlNode' && node.tagName === 'columnDef'
  );
  const newAugmented: ViewSets = {
    ...augmented,
    views: [
      {
        ...augmented.views[0],
        altViews: {
          ...augmented.views[0].altViews,
          altViews: [
            augmented.views[0].altViews.altViews[0],
            ...replaceItem(augmented.views[0].altViews.altViews, 1, {
              ...augmented.views[0].altViews.altViews[1],
              name: 'Accession View (Edited)',
            }),
          ],
        },
      },
    ],
    viewDefs: [
      {
        ...augmented.viewDefs[0],
        raw: {
          ...augmented.viewDefs[0].raw,
          children: [
            ...augmented.viewDefs[0].raw.children.slice(
              0,
              columnDefinitionIndex + 2
            ),
            {
              type: 'XmlNode',
              tagName: 'columnDef',
              attributes: {
                test: 'a',
              },
              children: [{ type: 'Text', string: 'test' }],
            },
            ...augmented.viewDefs[0].raw.children.slice(
              columnDefinitionIndex - 1
            ),
          ],
        },
      },
    ],
  };

  // Verify that can turn it all back
  expect(updateXml(deserializer(newAugmented))).toMatchSnapshot();
});
