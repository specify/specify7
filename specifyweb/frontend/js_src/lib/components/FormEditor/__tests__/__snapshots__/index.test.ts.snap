// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can edit a form definition 1`] = `
{
  "name": "Fish Views",
  "viewDefs": [
    {
      "legacyEditableDialog": true,
      "legacyGetTable": "edu.ku.brc.af.ui.forms.DataGetterForObj",
      "legacySetTable": "edu.ku.brc.af.ui.forms.DataSetterForObj",
      "name": "Accession",
      "table": "[table Accession]",
      "type": "form",
    },
  ],
  "views": [
    {
      "altViews": {
        "altViews": [
          {
            "default": true,
            "mode": "view",
            "name": "Accession View",
            "viewDef": "Accession",
          },
          {
            "default": false,
            "mode": "edit",
            "name": "Accession Edit",
            "viewDef": "Accession",
          },
        ],
      },
      "businessRules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "description": "The Accession form.",
      "legacyResourceLabels": false,
      "legacyUseBusinessRules": true,
      "name": "Accession",
      "table": "[table Accession]",
    },
  ],
}
`;

exports[`Can edit a form definition 2`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<viewset name=\\"Fish Views\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\">
	<views>
		<view name=\\"Accession\\" class=\\"edu.ku.brc.specify.datamodel.Accession\\" busrules=\\"edu.ku.brc.specify.datamodel.busrules.AccessionBusRules\\" resourcelabels=\\"false\\">
			<desc>The Accession form.</desc>
			<!-- Some comment -->
			<altviews>
				<altview name=\\"Accession View\\" viewdef=\\"Accession\\" mode=\\"view\\" default=\\"true\\"/>
				<altview name=\\"Accession View (2)\\" viewdef=\\"Accession\\" mode=\\"view\\" default=\\"true\\"/>
				<altview name=\\"Accession View (Edited)\\" viewdef=\\"Accession\\" mode=\\"edit\\"/>
			</altviews>
		</view>
	</views>
	<viewdefs>
		<viewdef type=\\"form\\" name=\\"Accession\\" class=\\"edu.ku.brc.specify.datamodel.Accession\\" gettable=\\"edu.ku.brc.af.ui.forms.DataGetterForObj\\" settable=\\"edu.ku.brc.af.ui.forms.DataSetterForObj\\">
			<desc>The Accession form.</desc>
			<enableRules/>
			<columnDef>100px,2px,175px,5px,140px,2px,161px,5px,75px,2px,160px,0px,15px,p:g</columnDef>
			<columnDef test=\\"a\\">test</columnDef>
			<columnDef>100px,2px,175px,5px,140px,2px,161px,5px,75px,2px,160px,0px,15px,p:g</columnDef>
			<columnDef os=\\"lnx\\">115px,2px,185px,5px,160px,2px,180px,5px,105px,2px,156px,0px,15px,p:g</columnDef>
			<columnDef os=\\"mac\\">130px,2px,205px,5px,180px,2px,200px,5px,125px,2px,176px,0px,15px,p:g</columnDef>
			<columnDef os=\\"exp\\">p,2px,p:g,5px:g,p,2px,p:g,5px:g,p,2px,p:g,0px,p,p:g</columnDef>
			<rowDef auto=\\"true\\" cell=\\"p\\" sep=\\"1px\\"/>
			<rows>
				<row>
					<!-- <cell type=\\"label\\" labelfor=\\"1\\"/> -->
					<cell type=\\"field\\" id=\\"1\\" name=\\"accessionNumber\\" uitype=\\"formattedtext\\"/>
					<cell type=\\"label\\" labelfor=\\"2\\"/>
					<cell type=\\"field\\" id=\\"2\\" name=\\"status\\" uitype=\\"combobox\\"/>
					<cell type=\\"label\\" labelfor=\\"3\\"/>
					<cell type=\\"field\\" id=\\"3\\" name=\\"type\\" uitype=\\"combobox\\"/>
				</row>
				<row>
					<cell type=\\"label\\" labelfor=\\"12\\"/>
					<cell type=\\"field\\" id=\\"12\\" name=\\"dateAccessioned\\" uitype=\\"formattedtext\\" uifieldformatter=\\"Date\\" default=\\"today\\"/>
					<cell type=\\"label\\" labelfor=\\"6\\"/>
					<cell type=\\"field\\" id=\\"6\\" name=\\"dateReceived\\" uitype=\\"formattedtext\\" uifieldformatter=\\"Date\\" default=\\"today\\"/>
				</row>
			</rows>
		</viewdef>
	</viewdefs>
</viewset>"
`;
